hack/tools/logcheck/main.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"go/ast"
+
+	"golang.org/x/tools/go/analysis/singlechecker"
+	"golang.org/x/tools/go/analysis"
+)
+
+// Doc explaining the tool.
+const Doc = "Tool to check use of unstructured logging patterns."
+
+// Analyzer runs static analysis.
+var Analyzer = &analysis.Analyzer{
+	Name: "logcheck",
+	Doc:  Doc,
+	Run:  run,
+}
+
+func main() {
+        singlechecker.Main(Analyzer)
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+
+	for _, file := range pass.Files {
+
+		ast.Inspect(file, func(n ast.Node) bool {
+
+			if fexpr, ok := n.(*ast.CallExpr); ok {
+				//fmt.Println(fexpr.Fun)
+				checkForFunctionExpr(fexpr.Fun, pass)
+			}
+
+			return true
+		})
+	}
+	return nil, nil
+}
+
+// checkForFunctionExpr checks for unstructured logging function, prints error if found any.
+func checkForFunctionExpr(fun ast.Expr, pass *analysis.Pass) {
+
+	if selExpr, ok := fun.(*ast.SelectorExpr); ok {
+		fName := selExpr.Sel.Name
```

Comment:
This tool lgtm! 

Can we also add some comments here on how we are parsing the klog functions? It took me a while to walk through the code and understand how `SelectorExpr` would behave for different unstructured logging cases. If we could add a comment here with one of the test cases as an example, that would be great :pray:

Reply:
Added :)